File: gfg/2024_7_july/python/05_vertical_height_of_binary_tree.py
# CODE

def verticalWidth(root):
    # base case
    if root is None: 
        return 0
    
    left = [0]
    right = [0]
    
    def inorder(root, pos, left, right):
        # base case
        if root is None:
            return
        
        # calling left
        inorder(root.left, pos - 1, left, right)
        
        # calculation
        if pos < left[0]:
            left[0] = pos
        elif pos > right[0]:
            right[0] = pos
        
        # calling right
        inorder(root.right, pos + 1, left, right)
    
    # calculate left and right in inorder function
    inorder(root, 0, left, right)
    
    return right[0] - left[0] + 1

File: gfg/2024_7_july/c++/05_vertical_height_of_binary_tree.cpp
// TIME COMPLEXITY
// O(n)

// SPACE COMPLEXITY
// O(n)

// CODE

class Solution {
  public:
    // Function to find the vertical width of a Binary Tree.
    void solve(Node *root, int &mini, int &maxi, int posi){
        if(root == NULL) return ;
        mini = min(mini, posi);
        maxi = max(maxi, posi);
        
        solve(root->left,mini,maxi,posi - 1);
        solve(root->right,mini,maxi,posi + 1);
    }
    int verticalWidth(Node* root) {
        if(root == NULL) return 0;
        int mini = 0;
        int maxi = 0;
        solve(root,mini,maxi,0);
        
        return maxi + abs(mini) + 1;
    }
};